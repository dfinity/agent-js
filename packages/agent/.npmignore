# We work with a safelist here, so block everything that's not permitted, and add packages
# that are.
**

!package.json
!README.md
!LICENSE

!src/**/*.ts
src/**/*.test.ts
src/**/__certificates__
src/package.json

!lib/**
lib/**/*.test.*
lib/**/*.tsbuildinfo
 