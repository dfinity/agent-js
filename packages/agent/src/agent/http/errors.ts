import { AgentError } from '../../errors';
import { HttpDetailsResponse } from '../api';

export class AgentHTTPResponseError extends AgentError {
  constructor(message: string, public readonly response: HttpDetailsResponse) {
    super(message);
    this.name = this.constructor.name;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
