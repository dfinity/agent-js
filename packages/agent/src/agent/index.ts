import { GlobalInternetComputer } from '../index';
import { Agent } from './api';

export * from './api';
export * from './http';
export * from './proxy';

declare const window: GlobalInternetComputer;
declare const global: GlobalInternetComputer;
declare const self: GlobalInternetComputer;

export function getDefaultAgent(): Agent {
  const agent =
    typeof window === 'undefined'
      ? typeof global === 'undefined'
        ? typeof self === 'undefined'
          ? undefined
          : self.ic.agent
        : global.ic.agent
      : window.ic.agent;

  if (!agent) {
    throw new Error('No Agent could be found.');
  }

  return agent;
}
