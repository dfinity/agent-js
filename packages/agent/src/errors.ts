/**
 * An error that happens in the Agent. This is the root of all errors and should be used
 * everywhere in the Agent code (this package).
 *
 * @todo https://github.com/dfinity/agent-js/issues/420
 */
export class AgentError extends Error {
  constructor(public readonly message: string) {
    super(message);
    Object.setPrototypeOf(this, AgentError.prototype);
  }
}
