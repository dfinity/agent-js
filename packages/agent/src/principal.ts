import { Buffer } from 'buffer/';
import { BinaryBlob, blobFromHex, blobFromUint8Array, blobToHex } from './types';
import { decode, encode } from './utils/base32';
import { getCrc32 } from './utils/getCrc';
import { sha224 } from './utils/sha224';

const SELF_AUTHENTICATING_SUFFIX = 2;
const ANONYMOUS_SUFFIX = 4;

export class Principal {
  public static anonymous(): Principal {
    return new this(blobFromUint8Array(new Uint8Array([ANONYMOUS_SUFFIX])));
  }

  public static selfAuthenticating(publicKey: BinaryBlob): Principal {
    const sha = sha224(publicKey);
    return new this(blobFromUint8Array(new Uint8Array([...sha, SELF_AUTHENTICATING_SUFFIX])));
  }

  public static from(other: unknown): Principal {
    if (typeof other === 'string') {
      return Principal.fromText(other);
    } else if (
      typeof other === 'object'
      && other !== null
      && (other as Principal)._isPrincipal === true
    ) {
      return new Principal((other as Principal)._blob);
    }

    throw new Error(`Impossible to convert ${JSON.stringify(other)} to Principal.`);
  }

  public static fromHex(hex: string): Principal {
    return new this(blobFromHex(hex));
  }

  public static fromText(text: string): Principal {
    const canisterIdNoDash = text.toLowerCase().replace(/-/g, '');

    let arr = decode(canisterIdNoDash);
    arr = arr.slice(4, arr.length);

    const principal = new this(blobFromUint8Array(arr));
    if (principal.toText() !== text) {
      throw new Error(`Principal "${principal.toText()}" does not have a valid checksum.`);
    }

    return principal;
  }

  public static fromBlob(blob: BinaryBlob): Principal {
    return new this(blob);
  }

  public readonly _isPrincipal = true;

  protected constructor(private _blob: BinaryBlob) {}

  public isAnonymous() {
    return this._blob.byteLength === 1 && this._blob[0] === ANONYMOUS_SUFFIX;
  }

  public toBlob(): BinaryBlob {
    return this._blob;
  }

  public toHash() {
    return this._blob;
  }

  public toHex(): string {
    return blobToHex(this._blob).toUpperCase();
  }

  public toText(): string {
    const checksumArrayBuf = new ArrayBuffer(4);
    const view = new DataView(checksumArrayBuf);
    view.setUint32(0, getCrc32(this.toBlob()), false);
    const checksum = Uint8Array.from(Buffer.from(checksumArrayBuf));

    const bytes = Uint8Array.from(this._blob);
    const array = new Uint8Array([...checksum, ...bytes]);

    const result = encode(array);
    const matches = result.match(/.{1,5}/g);
    if (!matches) {
      // This should only happen if there's no character, which is unreachable.
      throw new Error();
    }
    return matches.join('-');
  }

  public toString() {
    return this.toText();
  }
}
