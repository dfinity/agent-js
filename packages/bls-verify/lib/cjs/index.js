'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.blsVerify = void 0;
const bls12_381_1 = require('@noble/bls12-381');
const utils_1 = require('./utils');
/**
 * BLS Verification to be used in an IC Agent, using the @noble/bls-12-381 pure JS implementation
 * @param {Uint8Array | string} publicKey - Uint8Array or string of the public key used to verify a BLS signature
 * @param {Uint8Array | string} signature - digital signature
 * @param {Uint8Array | string} message - message to verify
 * @returns boolean
 */
const blsVerify = async (publicKey, signature, message) => {
  const pk = typeof publicKey === 'string' ? publicKey : (0, utils_1.toHex)(publicKey);
  const sig = typeof signature === 'string' ? signature : (0, utils_1.toHex)(signature);
  const msg = typeof message === 'string' ? message : (0, utils_1.toHex)(message);
  return await (0, bls12_381_1.verify)(sig, msg, pk);
};
exports.blsVerify = blsVerify;
//# sourceMappingURL=index.js.map
