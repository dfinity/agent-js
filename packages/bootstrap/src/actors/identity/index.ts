export { default as DocumentIdentities } from './DocumentIdentities';
export { default as IdentityActor } from './IdentityActor';
