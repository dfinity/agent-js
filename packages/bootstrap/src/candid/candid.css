:root {
    font-family: 'Circular Std', 'Roboto', sans-serif;
    font-weight: 300;
    font-size: 10px;
    --header-height: 3.8rem;
    --pad-sm: 0.8rem;
    --pad-md: 1.5rem;
    --pad-lg: 2.4rem;
    --font-sm: 1.2rem;
    --font-md: 1.4rem;
    --font-lg: 2.4rem;
    --white: #f9f9f9;
    --lightest: #ebf0fa;
    --lighter: #eaeaea;
    --light: #e5e5e5;
    --medium: #c3c3c4;
    --dark: #818284;
    --darker: #545454;
    --darkest: #292a2e;
    --error: #b82121;
}

#app {
    display: flex;
    flex-direction: column;
    width: 100vw;
    height: 100vh;
    font-size: var(--font-md);
}

#container {
    display: flex;
    flex-direction: column;
    height: calc(100vh - var(--header-height));
}

#main-content {
    flex: 1;
    overflow: auto;
}

#header {
    height: var(--header-height);
    width: 100%;
    text-align: center;
    font-size: var(--font-md);
    line-height: var(--header-height);
    color: var(--darkest);
    border-bottom: 1px solid var(--dark);
}

#title-card {
    padding: 3.2rem 1.8rem 0;
    font-size: 1.6rem;
}

#title {
    font-size: 3.4rem;
    letter-spacing: 0.1rem;
}

#methods {
    font-family: 'Vremena Grotesk', 'Roboto Mono', monospace;
    padding: 0;
    margin: 0;
}

#methods li {
    display: flex;
    flex-direction: column;
    list-style: none;
    padding: var(--pad-md);
    margin-top: var(--pad-lg);
    border: 1px solid var(--dark);
}

.signature {
    font-weight: lighter;
    margin-bottom: var(--pad-md);
}

.status,
.composite {
    margin-bottom: var(--pad-md);
}

input,
select {
    font-size: var(--font-md);
    font-family: 'Circular Std', 'Roboto', sans-serif;
    margin-bottom: var(--pad-lg);
}
input:not([type=checkbox]),
select {
    width: 100%;
    height: 4.8rem;
    padding: 1.5rem;
    border: 1px solid var(--dark);
}
input[type=number]::-webkit-inner-spin-button {
    width: 3.0rem;
    height: 3.0rem;
}
label {
    font-family: 'Circular Std', 'Roboto', sans-serif;
    margin-right: var(--pad-sm);
}
label.small {
    font-size: 1rem;
    text-transform: capitalize;
}
.popup-form {
    padding-left: var(--pad-md);
    width: 100%;
}
input[type=checkbox] ~ .popup-form {
    padding-left: var(--pad-md);
}
.input-container > div > .popup-form {
    padding-left: 0
}

.reject {
    border-color: var(--error);
}
.result {
    display: none;
    width: 100%;
    margin-top: 2.4rem;
    padding-top: 1.5rem;
    font-size: 1.4rem;
    border-top: 1px solid var(--light);
}
.error {
    color: var(--error);
}
.status {
    display: none;
    padding-top: 2.4rem;
    font-size: 1.2rem;
    color: var(--error);
    border-top: 1px solid var(--light);
}
.left {
    flex: 1;
    overflow-wrap: anywhere;
}
.right {
    color: var(--dark);
    align-self: stretch;
    margin-left: var(--pad-md);
}
.clear {
    clear: both;
}

.button-container {
    display: flex;
    margin-top: 0.8rem;
}

.btn {
    font-family: 'Roboto', sans-serif;
    letter-spacing: 0.05rem;
    background-color: var(--darkest);
    color: var(--lightest);
    font-size: 1.4rem;
    padding: 1.5rem;
    border: none;
    text-transform: uppercase;
    flex: 1;
}
.btn:not(:last-child) {
    margin-right: var(--pad-md);
}
.btn:focus,
input:focus,
select:focus,
textarea:focus {
    outline: none;
}
.random {
    background-color: var(--lighter);
    color: var(--darkest);
}

#console {
    overflow: hidden;
    height: var(--header-height);
    background-color: var(--white);
    color: var(--darker);
    padding: 0 var(--pad-md);
    transition: height .2s ease-in-out;
}
#console.open {
    height: 40vh;
}

#console-header {
    height: var(--header-height);
    padding: 0.8rem 0;
    border-bottom: 1px solid var(--light);
}
.collapse {
    display: inline-block;
    width: 2.2rem;
    height: 2.2rem;
    font-size: 1.8rem;
    padding-left: 0.5rem;
    margin-right: 0.5rem;
    line-height: 2rem;
    background-color: var(--medium);
    color: var(--white);
    border-radius: 50%;
    transform: rotate(90deg);
    transition: transform .2s ease;
}
#console.open .collapse {
    transform: rotate(-90deg);
}

#console-output {
    font-family: 'Vremena Grotesk', 'Roboto Mono', monospace;
    height: 0;
    padding: 0 0 3rem;
    margin-right: calc(-1 * var(--pad-md));
    overflow-y: auto;
}
#console.open #console-output {
    height: calc(40vh - var(--header-height));
}

.console-line {
    width: 100%;
    padding: 0.4rem 0.8rem;
    overflow-wrap: break-word;
}
.console-line:nth-child(2n-1) {
    background-color: #fff;
}

@media only screen and (min-width: 800px) {
    :root {
        --header-height: 4rem;
        --console-width: clamp(28rem, 30vw, 40rem);
    }
    
    #header {
        text-align: left;
        padding-left: 3rem;
    }
    
    #container {
        flex-direction: row;
        overflow: auto;
    }
    
    #title-card {
        padding-left: 0;
    }
    
    #main-content {
        flex: 1;
        padding: 0 3rem;
    }
    
    .btn {
        width: auto;
        flex: 0;
        padding: var(--pad-sm) var(--pad-md);
    }
    
    input:not([type=checkbox]),
    select {
        width: 23.2rem;
        height: 4rem;
        padding: 1rem var(--pad-md);
    }
    #console {
        display: flex;
        flex-direction: row-reverse;
        align-items: stretch;
        width: var(--header-height);
        height: calc(100vh - var(--header-height));
        padding: 0;
        transition: width .2s ease-in-out;
    }
    #console.open {
        width: calc(var(--header-height) + var(--console-width));
        height: calc(100vh - var(--header-height));
    }
    
    #console-header {
        position: relative;
        height: 100%;
        width: var(--header-height);
        padding: 0.8rem;
        border-bottom: none;
        border-right: 1px solid var(--light);
    }
    
    .collapse {
        transform: rotate(0deg);
    }
    #console.open .collapse {
        transform: rotate(180deg);
    }
    
    .console-label {
        white-space: nowrap;
        position: absolute;
        padding: 0.9rem clamp(10rem, 11vw, 40rem);
        background-color: var(--white);
        border-bottom: 1px solid var(--lighter);
        top: 0;
        right: var(--header-height);
    }
    
    #console-output {
        height: calc(100vh - (2 * var(--header-height)));
        width: 0;
        margin: var(--header-height) 0 0;
        transition: width .2s ease-in-out;
    }
    #console.open #console-output {
        height: calc(100vh - (2 * var(--header-height)));
        width: var(--console-width);
    }
}