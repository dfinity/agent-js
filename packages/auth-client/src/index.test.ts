import { Actor, HttpAgent } from '@dfinity/agent';
import { AgentError } from '@dfinity/agent/lib/cjs/errors';
import { IDL } from '@dfinity/candid';
import { Ed25519KeyIdentity } from '@dfinity/identity';
import { Principal } from '@dfinity/principal';
import { AuthClient, AuthClientStorage, ERROR_USER_INTERRUPT } from './index';

/**
 * A class for mocking the IDP service.
 */
class IdpMock {
  constructor(
    private readonly eventListener: (event: unknown) => void,
    private readonly origin: string,
  ) {}

  ready(origin?: string) {
    this.send(
      {
        kind: 'authorize-ready',
      },
      origin,
    );
  }

  send(message: unknown, origin?: string) {
    this.eventListener({
      origin: origin ?? this.origin,
      data: message,
    });
  }
}

const { location, fetch } = window;

beforeEach(() => {
  jest.useFakeTimers();
});
afterEach(() => {
  delete (window as any).location;
  (window as any).location = location;
});

describe('Auth Client', () => {
  it('should initialize with an AnonymousIdentity', async () => {
    const test = await AuthClient.create();
    expect(await test.isAuthenticated()).toBe(false);
    expect(test.getIdentity().getPrincipal().isAnonymous()).toBe(true);
  });

  it('should initialize with a provided identity', async () => {
    const identity = Ed25519KeyIdentity.generate();
    const test = await AuthClient.create({
      identity,
    });
    expect(test.getIdentity().getPrincipal().isAnonymous()).toBe(false);
    expect(test.getIdentity()).toBe(identity);
  });

  it('should log users out', async () => {
    const test = await AuthClient.create();
    await test.logout();
    expect(await test.isAuthenticated()).toBe(false);
    expect(test.getIdentity().getPrincipal().isAnonymous()).toBe(true);
  });
  it('should initialize an idleManager', async () => {
    const test = await AuthClient.create();
    expect(test.idleManager).toBeDefined();
  });
  it('should be able to invalidate an identity after going idle', async () => {
    // setup actor
    const identity = Ed25519KeyIdentity.generate();
    const mockFetch: jest.Mock = jest.fn();
    // http agent uses identity

    const canisterId = Principal.fromText('2chl6-4hpzw-vqaaa-aaaaa-c');
    const actorInterface = () => {
      return IDL.Service({
        greet: IDL.Func([IDL.Text], [IDL.Text]),
      });
    };

    // setup auth client
    const test = await AuthClient.create({
      identity,
      idleOptions: {
        idleTimeout: 1000,
      },
    });
    const httpAgent = new HttpAgent({ fetch: mockFetch });
    const actor = Actor.createActor(actorInterface, { canisterId, agent: httpAgent });

    test.idleManager?.registerCallback(() => {
      Actor.agentOf(actor)?.invalidateIdentity?.();
    });

    // wait for the idle timeout
    jest.advanceTimersByTime(1000);

    // check that the registered actor has been invalidated
    const expectedError =
      "This identity has expired due this application's security policy. Please refresh your authentication.";
    try {
      await actor.greet('hello');
    } catch (error) {
      expect((error as AgentError).message).toBe(expectedError);
    }
  });
  it('should log out after idle and reload the window by default', async () => {
    setup({
      onAuthRequest: () => {
        // Send a valid request.
        idpMock.send({
          kind: 'authorize-client-success',
          delegations: [
            {
              delegation: {
                pubkey: Uint8Array.from([]),
                expiration: BigInt(0),
              },
              signature: Uint8Array.from([]),
            },
          ],
          userPublicKey: Uint8Array.from([]),
        });
      },
    });
    delete (window as any).location;
    (window as any).location = { reload: jest.fn(), fetch };
    const mockFetch: jest.Mock = jest.fn();

    const canisterId = Principal.fromText('2chl6-4hpzw-vqaaa-aaaaa-c');
    const actorInterface = () => {
      return IDL.Service({
        greet: IDL.Func([IDL.Text], [IDL.Text]),
      });
    };

    const storage: AuthClientStorage = {
      remove: jest.fn(),
      get: jest.fn(),
      set: jest.fn(),
    };

    // setup auth client
    const test = await AuthClient.create({
      storage,
      idleOptions: {
        idleTimeout: 1000,
      },
    });

    // Test login flow
    await test.login({ identityProvider: 'http://localhost' });

    expect(storage.set).toBeCalled();
    expect(storage.remove).not.toBeCalled();

    const httpAgent = new HttpAgent({ fetch: mockFetch, host: 'http://127.0.0.1:8000' });
    const actor = Actor.createActor(actorInterface, { canisterId, agent: httpAgent });

    // simulate user being inactive for 10 minutes
    jest.advanceTimersByTime(10 * 60 * 1000);

    // Storage should be cleared by default after logging out
    expect(storage.remove).toBeCalled();

    expect(window.location.reload).toBeCalled();
  });
  it('should not reload the page if the default callback is disabled', async () => {
    setup({
      onAuthRequest: () => {
        // Send a valid request.
        idpMock.send({
          kind: 'authorize-client-success',
          delegations: [
            {
              delegation: {
                pubkey: Uint8Array.from([]),
                expiration: BigInt(0),
              },
              signature: Uint8Array.from([]),
            },
          ],
          userPublicKey: Uint8Array.from([]),
        });
      },
    });
    delete (window as any).location;
    (window as any).location = { reload: jest.fn(), fetch };

    const storage: AuthClientStorage = {
      remove: jest.fn(),
      get: jest.fn(),
      set: jest.fn(),
    };

    const test = await AuthClient.create({
      storage,
      idleOptions: {
        idleTimeout: 1000,
        disableDefaultIdleCallback: true,
      },
    });

    // Test login flow
    await test.login({ identityProvider: 'http://localhost' });

    expect(storage.set).toBeCalled();
    expect(storage.remove).not.toBeCalled();

    // simulate user being inactive for 10 minutes
    jest.advanceTimersByTime(10 * 60 * 1000);

    // Storage should not be cleared
    expect(storage.remove).not.toBeCalled();
    // Page should not be reloaded
    expect(window.location.reload).not.toBeCalled();
  });
  it('should not reload the page if a callback is provided', async () => {
    delete (window as any).location;
    (window as any).location = { reload: jest.fn(), fetch };
    const idleCb = jest.fn();
    const test = await AuthClient.create({
      idleOptions: {
        idleTimeout: 1000,
        onIdle: idleCb,
      },
    });

    // simulate user being inactive for 10 minutes
    jest.advanceTimersByTime(10 * 60 * 1000);

    expect(window.location.reload).not.toBeCalled();
    expect(idleCb).toBeCalled();
  });

  /**
   * This test reflects a feature that may be added at a future date,
   * allowing the authClient to register actors for automatic invalidation
   * and revalidation of identities
   */
  it.skip('should allow a registeredActor to get refreshed', async () => {
    setup({
      onAuthRequest: () => {
        // Send a valid request.
        idpMock.send({
          kind: 'authorize-client-success',
          delegations: [
            {
              delegation: {
                pubkey: Uint8Array.from([]),
                expiration: BigInt(0),
              },
              signature: Uint8Array.from([]),
            },
          ],
          userPublicKey: Uint8Array.from([]),
        });
      },
    });

    // setup actor
    const identity = Ed25519KeyIdentity.generate();
    const mockFetch: jest.Mock = jest.fn();
    // http agent uses identity

    const canisterId = Principal.fromText('2chl6-4hpzw-vqaaa-aaaaa-c');
    const actorInterface = () => {
      return IDL.Service({
        greet: IDL.Func([IDL.Text], [IDL.Text]),
      });
    };

    // setup auth client
    const test = await AuthClient.create({
      identity,
      idleOptions: {
        idleTimeout: 1000,
      },
    });
    const httpAgent = new HttpAgent({ fetch: mockFetch });
    const actor = Actor.createActor(actorInterface, { canisterId, agent: httpAgent });

    Actor.agentOf(actor)?.invalidateIdentity?.();

    // check that the registered actor has been invalidated
    const expectedError =
      "This identity has expired due this application's security policy. Please refresh your authentication.";
    try {
      await actor.greet('hello');
    } catch (error) {
      expect((error as AgentError).message).toBe(expectedError);
    }

    idpMock.ready();
    // check that the registered actor has been invalidated
    expect((Actor.agentOf(actor) as any)._identity).toBeTruthy();
  });
  it('should not set up an idle timer if the disable option is set', async () => {
    const idleFn = jest.fn();
    const test = await AuthClient.create({
      idleOptions: {
        idleTimeout: 1000,
        disableIdle: true,
      },
    });
    expect(idleFn).not.toHaveBeenCalled();
    expect(test.idleManager).toBeUndefined();
    // wait for default 30 minute idle timeout
    jest.advanceTimersByTime(30 * 60 * 1000);
    expect(idleFn).not.toHaveBeenCalled();
  });
});

// A minimal interface of our interactions with the Window object of the IDP.
interface IdpWindow {
  postMessage(message: { kind: string }): void;
  close(): void;
  closed: boolean;
}

let idpWindow: IdpWindow;
let idpMock: IdpMock;
function setup(options?: { onAuthRequest?: () => void }) {
  // Set the event handler.
  global.addEventListener = jest.fn((_, callback) => {
    // eslint-disable-next-line
    // @ts-ignore
    idpMock = new IdpMock(callback, 'https://identity.ic0.app');
  });

  // Mock window.open and window.postMessage since we can't open windows here.
  // eslint-disable-next-line
  // @ts-ignore
  global.open = jest.fn(() => {
    const idpWin: IdpWindow = (idpWindow = {
      postMessage: jest.fn(message => {
        if (message.kind === 'authorize-client') {
          options?.onAuthRequest?.();
        }
      }),
      close: jest.fn(() => {
        idpWin.closed = true;
      }),
      closed: false,
    });
    return idpWin;
  });
}

describe('Auth Client login', () => {
  it('should open a window with the IDP url', async () => {
    setup();
    const client = await AuthClient.create();
    // Try without #authorize hash.
    await client.login({ identityProvider: 'http://localhost' });
    expect(global.open).toBeCalledWith('http://localhost/#authorize', 'idpWindow', undefined);

    // Try with #authorize hash.
    global.open = jest.fn();
    await client.login({ identityProvider: 'http://localhost#authorize' });
    expect(global.open).toBeCalledWith('http://localhost/#authorize', 'idpWindow', undefined);

    // Default url
    global.open = jest.fn();
    await client.login();
    expect(global.open).toBeCalledWith(
      'https://identity.ic0.app/#authorize',
      'idpWindow',
      undefined,
    );

    // Default custom window.open feature
    global.open = jest.fn();
    await client.login({
      windowOpenerFeatures: 'toolbar=0,location=0,menubar=0',
    });
    expect(global.open).toBeCalledWith(
      'https://identity.ic0.app/#authorize',
      'idpWindow',
      'toolbar=0,location=0,menubar=0',
    );
  });

  it('should ignore authorize-ready events with bad origin', async () => {
    setup();
    const client = await AuthClient.create();
    await client.login();

    // Send an authorize-ready message with a bad origin. It should _not_ result
    // in a message sent back to the IDP.
    idpMock.ready('bad origin');

    // No response to the IDP canister.
    expect(idpWindow.postMessage).not.toBeCalled();
  });

  it('should respond to authorize-ready events with correct origin', async () => {
    setup();
    const client = await AuthClient.create();
    await client.login();

    // Send an authorize-ready message with the correct origin.
    idpMock.ready();

    // A response should be sent to the IDP.
    expect(idpWindow.postMessage).toBeCalled();
  });

  it('should call onError and close the IDP window on failure', async () => {
    setup({
      onAuthRequest: () => {
        // Send a failure message.
        idpMock.send({
          kind: 'authorize-client-failure',
          text: 'mock error message',
        });
      },
    });
    const client = await AuthClient.create();
    const failureFunc = jest.fn();
    await client.login({ onError: failureFunc });

    idpMock.ready();

    expect(failureFunc).toBeCalledWith('mock error message');
    expect(idpWindow.close).toBeCalled();
  });

  it('should call onError if recieved an invalid success message', async () => {
    setup({
      onAuthRequest: () => {
        idpMock.send({
          kind: 'authorize-client-success',
        });
      },
    });
    const client = await AuthClient.create();
    const failureFunc = jest.fn();
    await client.login({ onError: failureFunc });

    idpMock.ready();

    expect(failureFunc).toBeCalled();
    expect(idpWindow.close).toBeCalled();
    client.logout();
  });

  it('should call onError in an async pattern', async () => {
    setup({
      onAuthRequest: () => {
        idpMock.send({
          kind: 'authorize-client-success',
        });
      },
    });
    const client = await AuthClient.create();
    const cb = jest.fn();
    const failureFunc = async () => {
      await cb();
    };
    await client.login({ onError: failureFunc });

    idpMock.ready();

    expect(cb).toBeCalled();
    expect(idpWindow.close).toBeCalled();
  });

  it('should call onSuccess if recieved a valid success message', async () => {
    setup({
      onAuthRequest: () => {
        // Send a valid request.
        idpMock.send({
          kind: 'authorize-client-success',
          delegations: [
            {
              delegation: {
                pubkey: Uint8Array.from([]),
                expiration: BigInt(0),
              },
              signature: Uint8Array.from([]),
            },
          ],
          userPublicKey: Uint8Array.from([]),
        });
      },
    });

    const client = await AuthClient.create();
    const onSuccess = jest.fn();
    await client.login({ onSuccess: onSuccess });

    idpMock.ready();

    expect(onSuccess).toBeCalled();
    expect(idpWindow.close).toBeCalled();
    await client.logout();
  });

  it('should call onError if the user closed the IDP window', async () => {
    setup();
    jest.useRealTimers();
    const client = await AuthClient.create({ idleOptions: { disableIdle: true } });

    await expect(
      new Promise<void>((onSuccess, onError) =>
        (async () => {
          await client.login({ onSuccess, onError });
          idpWindow.close();
        })(),
      ),
    ).rejects.toMatch(ERROR_USER_INTERRUPT);
  });
  it('should call an async onSuccess if recieved a valid success message', async () => {
    setup({
      onAuthRequest: () => {
        // Send a valid request.
        idpMock.send({
          kind: 'authorize-client-success',
          delegations: [
            {
              delegation: {
                pubkey: Uint8Array.from([]),
                expiration: BigInt(0),
              },
              signature: Uint8Array.from([]),
            },
          ],
          userPublicKey: Uint8Array.from([]),
        });
      },
    });

    const client = await AuthClient.create();
    const cb = jest.fn();
    const onSuccess = async () => {
      cb();
    };
    await client.login({ onSuccess: onSuccess });

    idpMock.ready();

    expect(cb).toBeCalled();
    expect(idpWindow.close).toBeCalled();
  });
});
