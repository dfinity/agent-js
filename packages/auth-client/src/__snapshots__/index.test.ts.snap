// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Migration from Ed25519Key should continue using an existing Ed25519Key and delegation 1`] = `
DelegationIdentity {
  "_delegation": Object {
    "delegations": Array [
      Object {
        "delegation": Object {
          "expiration": "15e6375edc280000",
          "pubkey": "302a300506032b6570032100d1fa89134802051c8b5d4e53c08b87381b87097bca4c4f348611eb8ce6c91809",
        },
        "signature": "4e1365d5be0d4d0991088894ec850b6b1ffe7f1f00e53d82d6687f1aa8365f42129c2fee35f2f9d41c5749534b81017000c1a93095e406ff0ada03ed864d000a",
      },
    ],
    "publicKey": "302a300506032b6570032100d1fa89134802051c8b5d4e53c08b87381b87097bca4c4f348611eb8ce6c91809",
  },
  "_inner": Array [
    "302a300506032b6570032100d1fa89134802051c8b5d4e53c08b87381b87097bca4c4f348611eb8ce6c91809",
    "4bbff6b476463558d7be318aa342d1a97778d70833038680187950e9e02486c0d1fa89134802051c8b5d4e53c08b87381b87097bca4c4f348611eb8ce6c91809",
  ],
}
`;

exports[`Migration from Ed25519Key should generate and store a ECDSAKey if no key is stored and keyType is set to Ed25519 1`] = `"[\\"302a300506032b6570032100d1fa89134802051c8b5d4e53c08b87381b87097bca4c4f348611eb8ce6c91809\\",\\"4bbff6b476463558d7be318aa342d1a97778d70833038680187950e9e02486c0d1fa89134802051c8b5d4e53c08b87381b87097bca4c4f348611eb8ce6c91809\\"]"`;
